File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/TypedValueDto.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.taktik.icure.constants.TypedValuesType;

import java.io.Serializable;
import java.util.Date;

public class TypedValueDto implements Serializable, Comparable<TypedValueDto> {
	private static final long serialVersionUID = 1L;

	private TypedValuesType type;

	private Boolean booleanValue;
	private Integer integerValue;
	private Double doubleValue;
	private String stringValue;
	private Date dateValue;

	public TypedValueDto() {
	}

	public <T> TypedValueDto(T value) {
		setValue(value);
	}


	@SuppressWarnings("unchecked")
    @JsonIgnore
	public <T> T getValue() {
		if (type == null) {
			return null;
		}

		switch (type) {
			case BOOLEAN :
				return (T) booleanValue;
			case INTEGER :
				return (T) integerValue;
			case DOUBLE :
				return (T) doubleValue;
			case STRING : case CLOB: case JSON:
				return (T) stringValue;
			case DATE :
				return (T) dateValue;
		}

		return null;
	}

    @JsonIgnore
    public <T> void setValue(T value) {
		// Auto-detect type
		TypedValuesType type = null;
		if (value instanceof Boolean) {
			type = TypedValuesType.BOOLEAN;
		} else if (value instanceof Integer) {
			type = TypedValuesType.INTEGER;
		} else if (value instanceof Double) {
			type = TypedValuesType.DOUBLE;
		} else if (value instanceof String) {
			type = TypedValuesType.STRING;
		} else if (value instanceof Date) {
			type = TypedValuesType.DATE;
		}

		// Set type and value
		setTypeAndValue(type, value);
	}

	public <T> void setTypeAndValue(TypedValuesType type, T value) {
		// Set type
		this.type = type;

		// Reset value to null
		booleanValue = null;
		integerValue = null;
		doubleValue = null;
		stringValue = null;
		dateValue = null;

		// Set value if it matches the chosen type
		if (type != null && value != null) {
			switch (type) {
				case BOOLEAN :
					if (value instanceof Boolean) {
						booleanValue = (Boolean) value;
					}
					break;
				case INTEGER :
					if (value instanceof Integer) {
						integerValue = (Integer) value;
					}
					break;
				case DOUBLE :
					if (value instanceof Double) {
						doubleValue = (Double) value;
					}
					break;
				case STRING : case CLOB: case JSON:
					if (value instanceof String) {
						stringValue = (String) value;
					}
					break;
				case DATE :
					if (value instanceof Date) {
						dateValue = (Date) value;
					}
					break;
			}
		}
	}

	@SuppressWarnings({ "rawtypes", "unchecked" })
	@Override
	public int compareTo(TypedValueDto other) {
		return ((Comparable) getValue()).compareTo(other.getValue());
	}

	public TypedValuesType getType() {
		return type;
	}

	public void setType(TypedValuesType type) {
		this.type = type;
	}

	public void setValueType(TypedValuesType type) {
		this.type = type;
	}

	public Boolean getBooleanValue() {
		return booleanValue == null ? null : booleanValue == 1;
	}

	public void setBooleanValue(Boolean booleanValue) {
		this.booleanValue = booleanValue == null ? null : (booleanValue ? 1 : 0);
	}

	public Date getDateValue() {
		return dateValue;
	}

	public void setDateValue(Date dateValue) {
		this.dateValue = dateValue;
	}

	public Double getDoubleValue() {
		return doubleValue;
	}

	public void setDoubleValue(Double doubleValue) {
		this.doubleValue = doubleValue;
	}

	public Integer getIntegerValue() {
		return integerValue;
	}

	public void setIntegerValue(Integer integerValue) {
		this.integerValue = integerValue;
	}

	public String getStringValue() {
		return stringValue;
	}

	public void setStringValue(String stringValue) {
		this.stringValue = stringValue;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((booleanValue == null) ? 0 : booleanValue.hashCode());
		result = prime * result + ((dateValue == null) ? 0 : dateValue.hashCode());
		result = prime * result + ((doubleValue == null) ? 0 : doubleValue.hashCode());
		result = prime * result + ((integerValue == null) ? 0 : integerValue.hashCode());
		result = prime * result + ((stringValue == null) ? 0 : stringValue.hashCode());
		result = prime * result + ((type == null) ? 0 : type.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TypedValueDto other = (TypedValueDto) obj;
		if (booleanValue == null) {
			if (other.booleanValue != null)
				return false;
		} else if (!booleanValue.equals(other.booleanValue))
			return false;
		if (dateValue == null) {
			if (other.dateValue != null)
				return false;
		} else if (!dateValue.equals(other.dateValue))
			return false;
		if (doubleValue == null) {
			if (other.doubleValue != null)
				return false;
		} else if (!doubleValue.equals(other.doubleValue))
			return false;
		if (integerValue == null) {
			if (other.integerValue != null)
				return false;
		} else if (!integerValue.equals(other.integerValue))
			return false;
		if (stringValue == null) {
			if (other.stringValue != null)
				return false;
		} else if (!stringValue.equals(other.stringValue))
			return false;
		if (type != other.type)
			return false;
		return true;
	}
}

File: src/main/kotlin/org/taktik/icure/services/external/rest/v2/dto/embed/TypedValueDto.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.rest.v2.dto.embed

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.github.pozo.KotlinBuilder
import org.taktik.icure.constants.TypedValuesType
import org.taktik.icure.utils.InstantDeserializer
import org.taktik.icure.utils.InstantSerializer
import java.io.Serializable
import java.time.Instant
import java.util.*


@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class TypedValueDto<T>(
        val type: TypedValuesType? = null,
        val booleanValue: Boolean? = null,
        val integerValue: Int? = null,
        val doubleValue: Double? = null,
        val stringValue: String? = null,

        @JsonSerialize(using = InstantSerializer::class)
        @JsonInclude(JsonInclude.Include.NON_NULL)
        @JsonDeserialize(using = InstantDeserializer::class)
        val dateValue: Instant? = null,
        override val encryptedSelf: String? = null
) : Comparable<TypedValueDto<T>>, EncryptedDto, Serializable {
    companion object {
        fun <T> withValue(value: T): TypedValueDto<T> = withTypeAndValue(when (value) {
            is Boolean -> TypedValuesType.BOOLEAN
            is Int -> TypedValuesType.INTEGER
            is Double -> TypedValuesType.DOUBLE
            is String -> TypedValuesType.STRING
            is Date -> TypedValuesType.DATE
            else -> throw IllegalArgumentException("Unknown value type")
        }, value)

        fun <T> withTypeAndValue(type: TypedValuesType, value: T): TypedValueDto<T> = when (type) {
            TypedValuesType.BOOLEAN -> if (value is Boolean) {
                TypedValueDto(booleanValue = value, type = type)
            } else throw IllegalArgumentException("Illegal boolean value")
            TypedValuesType.INTEGER -> if (value is Int) {
                TypedValueDto(integerValue = value, type = type)
            } else throw IllegalArgumentException("Illegal integer value")
            TypedValuesType.DOUBLE -> if (value is Double) {
                TypedValueDto(doubleValue = value, type = type)
            } else throw IllegalArgumentException("Illegal double value")
            TypedValuesType.STRING, TypedValuesType.JSON, TypedValuesType.CLOB -> if (value is String) {
                TypedValueDto(stringValue = value, type = type)
            } else throw IllegalArgumentException("Illegal string value")
            TypedValuesType.DATE -> if (value is Instant) {
                TypedValueDto(dateValue = value, type = type)
            } else if (value is Date) {
                TypedValueDto(dateValue = (value as Date).toInstant(), type = type)
            } else throw IllegalArgumentException("Illegal date value")
        }
    }

    @JsonIgnore
    fun <T> getValue(): T? {
        if (type == null) {
            return null
        }
        return when (type) {
            TypedValuesType.BOOLEAN -> booleanValue as? T
            TypedValuesType.INTEGER -> integerValue as? T
            TypedValuesType.DOUBLE -> doubleValue as? T
            TypedValuesType.STRING, TypedValuesType.CLOB, TypedValuesType.JSON -> stringValue as? T
            TypedValuesType.DATE -> dateValue as? T
        }
    }

    override fun compareTo(other: TypedValueDto<T>): Int {
        return (other.getValue<T>() as Comparable<T>).compareTo(getValue<T>()!!)
    }

    override fun toString(): String {
        if (type != null) {
            when (type) {
                TypedValuesType.BOOLEAN -> return booleanValue.toString()
                TypedValuesType.INTEGER -> return integerValue.toString()
                TypedValuesType.DOUBLE -> return doubleValue.toString()
                TypedValuesType.STRING, TypedValuesType.CLOB, TypedValuesType.JSON -> return stringValue!!
                TypedValuesType.DATE -> return dateValue.toString()
            }
        }
        return super.toString()
    }
}
